### PR TITLE
Improve `--no-local` codegen

### PR DESCRIPTION
**Pull request by jabraham17**
Improves Chapel codegen with `--fast` and `--no-local` by always inlining `chpl_gen_comm_get` and `chpl_gen_comm_put` and by directly calling `memmove` (instead of through a wrapper.).

Prior to LLVM 16, we used the legacy LLVM pass manager which had a much more aggressive inliner, and these functions would be inlined into performance critical code. With newer LLVMs, the legacy pass manager has been replaced, and the new inliner is much less aggressive. When functions are inlined, the LLVM optimizer has more to work with and can optimize better and perform various forms of LICM, which is important for potentially expensive operations like GET/PUT.

Additionally, when accessing a wide pointer on the same node, the Chapel runtime uses `memmove`. LLVM has a number of optimizations for memory functions like `memmove` and `memcpy`. However, the Chapel runtime wraps `memmove` in `chpl_memmove` and if the inliner is not aggressive enough to inline the call, then the LLVM optimizations will fail to fire.

The particular patterns that cause these performance issues is exemplified by LCALs, a vectorization test suite. However, the Chapel implementation (and presumable the C version as well) uses various indirections to reduce boiler-plate. This means for a given array access, there may be several wide pointers to go through for a single array access. I created a reproducer of this pattern in `test/llvm/inlining/no-get-put.chpl`. If you remove the class and nested arrays, this code fully vectorizes loop in `kernel()`. However, the class and nested arrays both introduce indirection which is codegened as a wide pointer. This means the loop doesn't vectorize and may be extremely slow if the pointer accesses are not LICMed out of the core loop. I consider improving the vectorization in the presence of wide pointers future work.

In summary, if `chpl_gen_comm_get`, `chpl_gen_comm_put`, and `chpl_memmove` are not inlined properly, performance suffers with `--no-local`. A simple example is `muladdsub` from LCALs, which takes a 6x performance regression

Testing:
- [x] tested that `LCALs` performance recovers to pre-LLVM 16 levels with `--no-local`
- [x] full correctness test with/without comm

[Reviewed by @mppf]

### CODE REVIEW

File: runtime/include/chpl-comm-compiler-macros.h [hunk 1]
```diff
@@ -53,7 +53,7 @@ void chpl_gen_comm_get(void *addr, c_nodeid_t node, void* raddr,
                        size_t size, int32_t commID, int ln, int32_t fn)
```
Comment (mppf) [inline runtime/include/chpl-comm-compiler-macros.h:53]: Perhaps this could use `__attribute__((always_inline))` just before this declaration? Maybe put it in a macro & only use it for GCC and clang?

Our CHANGES.md says the runtime eliminated use of `__always_inline` in 1.12. Digging a bit, that happened in PR #1631  because `__always_inline` was always a workaround for problems building with GASNet, and we found a better solution.

The old way we defined it was

```
#ifdef __GNUC__
#define ___always_inline inline __attribute__((__always_inline__))
#else
#define ___always_inline inline
#endif
```

**Review by mppf — APPROVED** [review]
Please run the idea of this PR past Brad (at least) to make sure we feel OK about it going in at this point in the release cycle.

File: compiler/codegen/cg-expr.cpp [hunk 2]
```diff
@@ -3289,10 +3289,15 @@ GenRet codegenCallExprWithArgs(const char* fnName,
...
 #ifdef HAVE_LLVM
-    fn.val = getFunctionLLVM(fnName);
-    if (fn.val == nullptr) {
+    auto func = getFunctionLLVM(fnName);
+    if (func == nullptr) {
       INT_FATAL(fnSym, "unable to find function %s\n", fnName);
     }
+    if (fnName == astr_gen_comm_get || fnName == astr_gen_comm_put) {
+      // these are performance-critical functions that we want to inline
+      func->addFnAttr(llvm::Attribute::AlwaysInline);
```
[the remark this answers is not in the document]
Comment (jabraham17) [reply compiler/codegen/cg-expr.cpp:3298]: I agree this is not ideal, I tried adding this to `compiler/codegen/cg-symbol.cpp:2500` and did not get the same perf benefits. I will investigate using `always_inline` in the C code and see if thats better (I forgot that attribute existed)

[the remark this answers is not in the document]
Comment (mppf) [reply compiler/codegen/cg-expr.cpp:3298]: It's also fine with me to leave that as future work.

**Comment by jabraham17** [comment]
To close the loop on this, after offline discussions with Michael and Brad, I will merge this PR today so it goes into the release. I have opened https://github.com/chapel-lang/chapel/pull/25281 which I will rebase and merge after the release to incorporate Michaels feedback on this PR. This followup will wait until after the release as we have some portability concerns that should be tested more thoroughly. I have already validated using `___always_inline` will work the same as this PR, albeit in a cleaner way.
